### PR TITLE
Fix for Google Lighthouse / PageSpeed Insight Tool being shown a 406 status code

### PR DESCRIPTION
**Pull request by ankurp**
Fix for `allow_browser` throwing a 406 status code when request are made from Google Lighthouse / PageSpeed Insight Tool

### Motivation / Background

Google PageSpeed Insight is a tool that makes requests from a server to the website specificed. If it makes a request to newer Rails websites that use allow_browser feature then it returns a 406 status code as the browser is slightly older but we should treat this same as a bot like we do for crawling purpose and generate Page speed insights.

This Pull Request has been created to solve this issue where we want to support modern browser but also want bots to crawl without being thrown a 406 status code like in the case for Google's Lighthouse Tool.

image

### Detail

This PR bumps the useragent gem which has a fix for this where Lighthouse Tool is considered a bot and lets the request go through instead of terminating the request early with a 406 status code and new 406 page.

### Checklist

Before submitting the PR make sure the following are checked:

* [ ] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -19,3 +19,4 @@ pkg/
 /tmp/
 /yarn-error.log
 /test-reports/
+.DS_Store
```
Comment (simi) [inline .gitignore:22]: :scissors:

Comment (ankurp) [reply .gitignore]: Done. Is there a reason to not gitignore this?

Comment (bradleypriest) [reply .gitignore]: @ankurp this has been proposed a few times before, reasoning here https://github.com/rails/rails/pull/22696

**Comment by fatkodima** [comment]
Squash commits into 1.

**Comment by ankurp** [comment]
@fatkodima Done

**Comment by byroot** [comment]
I think adding bots to a list just can't scale. I really don't like user agent filtering as a concept in the first place, but if we're are to do it, maybe we should just show an error when we know the browser is too old, and let unknown browser through?

But I don't even know if that's possible given bots tend to somewhat pretend to be browsers for that exact reason.

@dhh since you added the user agent filtering as a default, what do you think should be the way forward here?

**Comment by dhh** [comment]
Yeah, don't want to play that game of chasing user agents. I think it's worth exploring a way to only gate on known browsers where the versions are too low. But not sure that would even help this case? Looks like Lighthouse identifies as Chrome 94?

**Comment by ankurp** [comment]
If we want to check for bots in rails framework rather than depend on useragent gem then it is a good approach as I feel the gem is not maintained.

**Comment by byroot** [comment]
> Looks like Lighthouse identifies as Chrome 94?

That's the problem with user agents, since in the past it was used to reject browsers, all browsers pretent to be someone else. So yeah Lighthouse give an user agent that's very much like Chrome 94 with just an added `Chrome-Lighthouse` at the end.

So I'm not even sure a it would be viable to only reject known incompatible browser.

I have zero context on why some browsers are rejected by default, I assume it's because hotwire require some features? I haven't done much frontend in a long time but back in the day what was considered best practice was to feature test client side. Isn't that possible with hotwire?

**Comment by dhh** [comment]
This is a catch-all that's provides the ground floor for all the features that both Hotwire and nobuild and import maps require. So that part isn't changing. You can always add a conditional to this so you can do ?sidestep=1 or whatever to route around it.

But we aren't going to fundamentally change how this works, no.

**Comment by dhh** [comment]
Although I guess I could see a backed in side door. So having allow_browser automatically recognize something ala ?allow_any_browser=1 as a query param that then skips the browser check.

**Comment by byroot** [comment]
> So that part isn't changing.

I'm not gonna insist because ultimately I don't really care about frontend, so I'll let you resolve that issue. I just find it weird we're having such issues when from what I understand all the browsers we're checking for all have auto-updates.

**Comment by ankurp** [comment]
Let me know how I should proceed with this. I can add a query param by pass but they seems more hacky and bots that crawl may nit be able to add that query param on all pages

**Comment by dhh** [comment]
I'd take a patch that turns off the check if ?allow_browser=1 is passed. That'll of course only work when you control the URL submitted (like with Lighthouse). If you can't control the URL, and you want your app crawled, you'll have to just not use this feature, and feature gate on your own.

**Comment by rafaelfranca** [comment]
Wait folks, I'm not sure why we are even discussing this PR. It doesn't even need to be merged. The issue was fixed by the release of [`useragent 0.16.11`](https://github.com/gshutler/useragent/compare/v0.16.10...v0.16.11). There is nothing special that Rails need to do for Lighthouse requests to passthrough. Applications that care about the problem just need to upgrade useragent.

This PR is mostly adding a test, that I don't think it is necessary.

For the general problem, if people care about crawlers, "you'll have to just not use this feature, and feature gate on your own" is a good enough solution to me.

Thank you for fixing the issue on useragent and this PR here, but it isn't necessary, so I'm going to close.

**Comment by rafaelfranca** [comment]
No need. This version is only used for development. Rails apps can use any version of useragent, there is no need to change the framework code

**Pull request closed without merging**